### PR TITLE
feature: Update Dynamic table parameters on materialization

### LINKED ISSUE #859: [ADAP-1058] [Feature] Implement refresh strategy for dynamic tables
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt-snowflake functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

Provide an option for analytic engineers to delay the refresh of data until the next scheduled refresh instead of forcing a refresh upon object creation. The primary benefit of this feature is that the entire pipeline can be built and the logic can be validated with needing to run a potentially expensive query to initially populate the dynamic table. The proposed name for this configuration is `refresh_strategy`, as identified in #603.

### Describe alternatives you've considered

_No response_

### Who will this benefit?

This will benefit analytics engineers that heavily use dynamic tables and who want to validate logic in their pipelines without reloading data. This also allows a little more flexibility and control over when data is populated. For example, the engineer may not want to run an expensive query in the middle of the day, but also may not want to wait until overnight to deploy the dynamic table.

### Are you interested in contributing this feature?

_No response_

### Anything else?

It's worth noting that Snowflake does not currently support this functionality, making this impossible. However, if Snowflake were to add this, then we can work that into `dbt-snowflake`. This issue is a placeholder for if/when that happens.

### LINKED ISSUE #868: [ADAP-1076] [Feature] [Dynamic Tables] Add new parameter properties (REFRESH_MODE, INITIALIZE)
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt-snowflake functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

New parameters have been added to dynamic tables, please add them so that users can add them to the config and use the parameters.
The parameters are REFRESH_MODE, INITIALIZE.
image

### Describe alternatives you've considered

_No response_

### Who will this benefit?

_No response_

### Are you interested in contributing this feature?

_No response_

### Anything else?

https://docs.snowflake.com/en/sql-reference/sql/create-dynamic-table

### PR DESCRIPTION
**Pull request by HenkvanDyk**
resolves #868 
[docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose) dbt-labs/docs.getdbt.com/#

Documentation updates here  https://github.com/dbt-labs/docs.getdbt.com/pull/4835
### Problem

Dynamic tables were refreshed on each run in 1.6.x, but not in 1.7.x which resulted in 2 things:
1. In 1.6.x the recreation of a dynamic table would trigger a refresh. Any downstream models which are recreated in the same run could (if the upstream target_lag = Downstream) trigger upstream models again. This resulted in redundant refreshes and delayed runs.
2. Users who upgraded to 1.7.x who actually want to refresh a dynamic table on a run would need to add a macro to trigger a refresh (I'd want to do this to ensure my orchestration system checks that a table is up to date before downstream tables read from it)

### Solution

Snowflake introduced new parameters which can be set on dynamic tables creation. [Docs ](https://docs.snowflake.com/en/sql-reference/sql/create-dynamic-table)
The `INITIALIZE` feature allows one to refresh a dynamic table either only on schedule, or also on creation.
This could allow orchestration of refreshes to be done when a dbt run is invoked, which "creates" the table again.
This could resolve #859 as well, as DBT would not need to set a refresh strategy.

Alterations of these parameters are not included as these settings cannot be altered after creation. https://docs.snowflake.com/en/sql-reference/sql/alter-dynamic-table#parameters

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [ ] I have run this code in development and it appears to resolve the stated issue
- [ ] This PR includes tests, or tests are not required/relevant for this PR
- [ ] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX

### CODE REVIEW

File: dbt/adapters/snowflake/relation_configs/dynamic_table.py [hunk 1]
```diff
@@ -71,6 +80,9 @@ def parse_relation_results(cls, relation_results: RelationResults) -> Dict:
             "query": dynamic_table.get("text"),
             "target_lag": dynamic_table.get("target_lag"),
             "snowflake_warehouse": dynamic_table.get("warehouse"),
+            "refresh_mode": dynamic_table.get("refresh_mode"), # Should I set defaults here or do they get passed as null
+            "initialize": dynamic_table.get("initialize"), # Should I set defaults here or do they get passed as null
```
Comment (HenkvanDyk) [inline dbt/adapters/snowflake/relation_configs/dynamic_table.py:84]:
```suggestion
            "refresh_mode": dynamic_table.get("refresh_mode"),
            "initialize": dynamic_table.get("initialize"),
```

**Comment by HenkvanDyk** [comment]
I am having a hard time testing this locally to check the creation implements the features as expected.

I've tried installing dbt-snowflake but when running tests the import is failing - any guidance on the structure of my venv would be appreciated.

image

**Comment by mikealfare** [comment]
Thanks for your contribution @HenkvanDyk! The implementation for object creation looks fine, but there are a few more items to address. At a high level:

- update the [describe query](https://github.com/dbt-labs/dbt-snowflake/blob/main/dbt/include/snowflake/macros/relations/dynamic_table/describe.sql) to pull back the new attributes
- incorporate those new attributes when producing the changeset [here](https://github.com/dbt-labs/dbt-snowflake/blob/4ff6bdb46b8d1a9282b475931ba753806f1eef0c/dbt/adapters/snowflake/relation.py#L56-L82) and [here](https://github.com/dbt-labs/dbt-snowflake/blob/4ff6bdb46b8d1a9282b475931ba753806f1eef0c/dbt/adapters/snowflake/relation_configs/dynamic_table.py#L79-L117)
- add/update [tests](https://github.com/dbt-labs/dbt-snowflake/tree/main/tests/functional/adapter/dynamic_table_tests)

Adding the new fields to the describe macro will allow for them to be checked in the tests; the tests call the describe macro to inspect the object in the database (which in turn also tests the describe macro if you provide a non-default setting).

**Comment by mikealfare** [comment]
> I am having a hard time testing this locally to check the creation implements the features as expected.
> 
> I've tried installing dbt-snowflake but when running tests the import is failing - any guidance on the structure of my venv would be appreciated.
> 
> image

This looks like it might be a result of our decoupling work earlier this year (the error is familiar). We pulled what we affectionately referred to as "the adapters zone" into its own package `dbt-adapters` [here](https://github.com/dbt-labs/dbt-adapters), and then removed `dbt-core` as a functional dependency within the adapter implementations (e.g. `dbt-snowflake`). Depending on when you pulled down these branches, you might have gotten something that was still under construction. That period has since passed though, so you should be able to setup a fresh environment now.

**Comment by mikealfare** [comment]
This PR will be continued in #1081. I'm closing this to avoid confusion, however the work will continue. Please refer to #1081 moving forward.

**Pull request closed without merging**